graphene/relay/connection.py

```diff
@@ -90,65 +90,57 @@ class ConnectionBase(AbstractType):
 
 
 class Connection(six.with_metaclass(ConnectionMeta, ObjectType)):
-    pass
-
-
-class IterableConnectionField(Field):
-
-    def __init__(self, type, *args, **kwargs):
-        kwargs.setdefault('before', String())
-        kwargs.setdefault('after', String())
-        kwargs.setdefault('first', Int())
-        kwargs.setdefault('last', Int())
-        super(IterableConnectionField, self).__init__(
-            type,
-            *args,
-            **kwargs
-        )
-
-    @property
-    def type(self):
-        type = super(IterableConnectionField, self).type
-        if is_node(type):
-            connection_type = type.Connection
-        else:
-            connection_type = type
-        assert issubclass(connection_type, Connection), (
-            '{} type have to be a subclass of Connection. Received "{}".'
-        ).format(str(self), connection_type)
-        return connection_type
 
     @classmethod
-    def resolve_connection(cls, connection_type, args, resolved):
-        if isinstance(resolved, connection_type):
-            return resolved
+    def Field(cls, *args, **kwargs):  # noqa: N802
+        return ConnectionField(cls, *args, **kwargs)
 
+    @classmethod
+    def connection_resolver(cls, resolved, args, context, info):
         assert isinstance(resolved, Iterable), (
             'Resolved value from the connection field have to be iterable or instance of {}. '
             'Received "{}"'
-        ).format(connection_type, resolved)
+        ).format(cls, resolved)
         connection = connection_from_list(
             resolved,
             args,
-            connection_type=connection_type,
-            edge_type=connection_type.Edge,
+            connection_type=cls,
+            edge_type=cls.Edge,
             pageinfo_type=PageInfo
         )
         connection.iterable = resolved
         return connection
 
+
+class ConnectionField(Field):
+
+    def __init__(self, type, *args, **kwargs):
+        kwargs.setdefault('before', String())
+        kwargs.setdefault('after', String())
+        kwargs.setdefault('first', Int())
+        kwargs.setdefault('last', Int())
+        assert issubclass(type, Connection), (
+            '{} type have to be a subclass of Connection. Received "{}".'
+        ).format(str(self), type)
+        super(ConnectionField, self).__init__(
+            type,
+            *args,
+            **kwargs
+        )
+
     @classmethod
     def connection_resolver(cls, resolver, connection_type, root, args, context, info):
         resolved = resolver(root, args, context, info)
 
-        on_resolve = partial(cls.resolve_connection, connection_type, args)
+        if isinstance(resolved, connection_type):
+            return resolved
+
+        on_resolve = partial(connection_type.connection_resolver, args=args, context=context, info=info)
         if isinstance(resolved, Promise):
```

Comment:
The only minor problem I have with this is that the instance check does not allow for duck-typing, which goes in line with https://github.com/graphql-python/graphql-core/pull/91 and is a common principle in python